File: CS50x/sentimental-credit/credit.py
def validCredit(number, z):

    i = z - 1
    a = 0
    sum = 0

    if (z % 2 == 0):
        while a < i:
            digit = int(number[a]) * 2
            if digit > 9:
                digit = digit - 9
            sum = sum + digit
            sum = sum + int(number[a + 1])
            a = a + 2

    else:
        sum = int(number[0])
        while (a + 1) < i:
            digit = int(number[a + 1]) * 2
            if digit > 9:
                digit = digit - 9
            sum = sum + digit
            sum = sum + int(number[a + 2])
            a = a + 2

    return sum

def creditFlag(number, z):

    fd = 10 * int(number[0]) + int(number[1])

    if ((fd == 34) or (fd == 37)) and (z == 15):
        print("AMEX\n")
    elif ((fd == 51) or (fd == 52) or (fd == 53) or (fd == 54) or (fd == 55)) and (z == 16):
        print("MASTERCARD\n")
    elif (int(number[0]) == 4) and ((z == 13) or (z == 16)):
        print("VISA\n")
    else:
        print("INVALID\n")


def main():
    number = input("Number: ")
    z = len(number)
    x = validCredit(number, z)

    if (x % 10) == 0:
        creditFlag(number, z)
    else:
        print("INVALID\n")


if __name__ == "__main__":
    main()

File: CS50x/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long long card_number;

    // Get user input for the credit card number
    do
    {
        card_number = get_long_long("Number: ");
    }
    while (card_number <= 0);

    // Count the number of digits in the card number
    int digit_count = 0;
    long long temp = card_number;
    while (temp > 0)
    {
        temp /= 10;
        digit_count++;
    }

    // Extract the first two digits of the card number
    long long first_two_digits = card_number;
    while (first_two_digits >= 100)
    {
        first_two_digits /= 10;
    }

    // Check the card type based on the first digits and the number of digits
    if ((digit_count == 15 && (first_two_digits == 34 || first_two_digits == 37)) ||
        (digit_count == 16 && (first_two_digits >= 51 && first_two_digits <= 55)) ||
        ((digit_count == 13 || digit_count == 16) && first_two_digits / 10 == 4))
    {
        // Perform Luhn's algorithm to validate the card number
        int sum = 0;
        bool multiply = false;
        while (card_number > 0)
        {
            int digit = card_number % 10;
            card_number /= 10;

            if (multiply)
            {
                digit *= 2;
                if (digit > 9)
                {
                    digit = digit % 10 + digit / 10;
                }
            }

            sum += digit;
            multiply = !multiply;
        }

        // Check if the card number is valid
        if (sum % 10 == 0)
        {
            if (digit_count == 15 && (first_two_digits == 34 || first_two_digits == 37))
            {
                printf("AMEX\n");
            }
            else if (digit_count == 16 && (first_two_digits >= 51 && first_two_digits <= 55))
            {
                printf("MASTERCARD\n");
            }
            else if ((digit_count == 13 || digit_count == 16) && first_two_digits / 10 == 4)
            {
                printf("VISA\n");
            }
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }

    return 0;
}
